src/rpc/network/socket_pool.h

```diff
@@ -0,0 +1,142 @@
+/*!
+ *  Copyright (c) 2021 by Contributors
+ * \file socket_pool.h
+ * \brief Socket pool of nonblocking sockets for DGL distributed training.
+ */
+#ifndef DGL_RPC_NETWORK_SOCKET_POOL_H_
+#define DGL_RPC_NETWORK_SOCKET_POOL_H_
+
+#ifdef USE_EPOLL
+
+// Currently only support linux operating system with epoll
+#include <sys/epoll.h>
+#include <unordered_map>
+#include <queue>
+#include <memory>
+
+namespace dgl {
+namespace network {
+
+class TCPSocket;
+
+/*!
+ * \brief SocketPool maintains a group of nonblocking sockets, and can provide
+ * active sockets.
+ * Currently SocketPool is based on epoll, a scalable I/O event notification
+ * mechanism in Linux operating system. 
+ */
+class SocketPool {
+ public:
+  /*!
+   * \brief SocketPool constructor
+   */
+  SocketPool();
+
+  /*!
+   * \brief Add a socket to SocketPool
+   * \param socket tcp socket to add
+   * \param socket_id receiver/sender id of the socket
+   * \param events EPOLLIN, EPOLLOUT, or EPOLLIN|EPOLLOUT
+   */
+  void AddSocket(std::shared_ptr<TCPSocket> socket, int socket_id,
+    unsigned int events = EPOLLIN);
+
+  /*!
+   * \brief Remove socket from SocketPool
+   * \param socket tcp socket to remove
+   * \return number of remaing sockets in the pool
+   */
+  size_t RemoveSocket(std::shared_ptr<TCPSocket> socket);
+
+  /*!
+   * \brief SocketPool destructor
+   */
+  ~SocketPool();
+
+  /*!
+   * \brief Get current active socket. This is a blocking method
+   * \param socket_id output parameter of the socket_id of active socket
+   * \return active TCPSocket
+   */
+  std::shared_ptr<TCPSocket> GetActiveSocket(int* socket_id);
+
+ private:
+  /*!
+   * \brief Wait for event notification
+   */
+  void Wait();
+
+  /*!
+   * \brief map from fd to TCPSocket
+   */
+  std::unordered_map<int, std::shared_ptr<TCPSocket>> tcp_sockets_;
+
+  /*!
+   * \brief map from fd to socket_id
+   */
+  std::unordered_map<int, int> socket_ids_;
+
+  /*!
+   * \brief fd for epoll base
+   */
+  int epfd_;
+
+  /*!
+   * \brief number of pending active events
+   */
+  int pending_events_;
+
+  /*!
+   * \brief index of current active event
+   */
+  int current_event_;
+
+  static const int MAX_EVENTS = 10;
+  /*!
+   * \brief array of epoll event
+   */
+  epoll_event events_[MAX_EVENTS];
+};
+
+}  // namespace network
+}  // namespace dgl
+
+#else
```

Comment:
It's hard to track the API if you break it in the header with different conditions. Could you try make a unified interface here and let the different implementation happens in the cc file?

Reply:
You can consider using template or something like`if (EPOLL_DEFINER) {...} else {...}`

Reply:
Sure